ultra/baselines/common/social_vehicle_extraction.py

```diff
@@ -0,0 +1,270 @@
+    if len(vehicles_on_ego_lane) > 0:
+        front_dist, rear_dist = float("inf"), float("inf")
+        front_vehicle, rear_vehicle = None, None
+        for vehicle in vehicles_on_ego_lane:
+            location, _ = find_relative_location(
+                vehicle.position, ego_vehicle_pos, ego_vehicle_heading
+            )
+            dist = get_distance(vehicle, ego_vehicle_pos)
+            if location == FRONT and dist < front_dist:
+                front_vehicle = vehicle
+                front_dist = dist
+            elif location == REAR and dist < rear_dist:
+                rear_vehicle = vehicle
+                rear_dist = dist
+        if front_vehicle:
+            social_vehicles.append(
+                extractor_func(
+                    social_vehicle=front_vehicle,
+                    ego_vehicle_pos=ego_vehicle_pos,
+                    ego_vehicle_heading=ego_vehicle_heading,
+                )
+            )
+        if rear_vehicle:
+            social_vehicles.append(
+                extractor_func(
+                    social_vehicle=rear_vehicle,
+                    ego_vehicle_pos=ego_vehicle_pos,
+                    ego_vehicle_heading=ego_vehicle_heading,
+                )
+            )
+
+    # process vehicles on the other lanes:
+    # sort vehicles based on their distance per lane and keep num_social_vehicle_per_lane
+    for lane, vehicles in vehicles_on_other_lanes.items():
+        vehicles_on_other_lanes[lane].sort(
+            key=lambda vehicle: get_distance(vehicle, ego_vehicle_pos)
+        )
+        vehicles_on_other_lanes[lane] = vehicles_on_other_lanes[lane][
+            : min(len(vehicles_on_other_lanes), num_social_vehicle_per_lane)
+        ]
+        for v in vehicles_on_other_lanes[lane]:
+            social_vehicles.append(
+                extractor_func(v, ego_vehicle_pos, ego_vehicle_heading)
+            )
+    # add fake social vehicles if necessary
+    if len(social_vehicles) < social_capacity:
+        # find a waypoint on ego path equal to max_dist
+        fake_social_vehicle = make_fake_social_vehicle(
+            ego_vehicle_pos, ego_vehicle_heading, prev_ego_wp, max_dist_social_vehicle
+        )
+        fake_social_vehicles = [
+            fake_social_vehicle for i in range(social_capacity - len(social_vehicles))
+        ]
+        social_vehicles.extend(fake_social_vehicles)
+
+    # Permute the social vehicles randomly
+    social_vehicles_permuted = np.random.permutation(social_vehicles)
+    # If we have more vehicles than required, then delete the tail
+    if len(social_vehicles_permuted) > social_capacity:
+        social_vehicles_permuted = social_vehicles_permuted[:social_capacity]
+
+    return social_vehicles_permuted
+
+
+def make_fake_social_vehicle(
+    ego_vehicle_pos, ego_vehicle_heading, prev_ego_wp, max_dist_social_vehicle
+):
+    # DG: I changed to zeros because I think it'll be easier for learning
+    return [0.0, 0.0, 0.0, 0.0]
```

Comment:
Is this the new default or will it be reverted to the commented version below?

Reply:
it's supposed to fill the empty spots for social vehicles, say if we have 20 vehicles as the input of the network but there are not enough vehicles in the range we fill them with 0